### PR TITLE
Dry run option for clean command

### LINKED ISSUE #5834: Make users aware of (the need for) datalad clean
Creating this issue to separate aspects of #5519.

Not only, but particularly the extraction cache of `datalad-archives` is a surprising and not easy to discover location requiring disc space. A `--dry-run` for `clean` would help discover it and in addition the special remote itself shouldn't be silent about it, but announce the creation of that extraction cache.

### PR DESCRIPTION
**Pull request by bpoldrack**
There are several places where temporary files could be left by datalad operations. `datalad clean` now comes with a `--dry-run` option to just report on what could be cleaned.
Furthermore, the `datalad-archives` special remote now issues an INFO message to annex when it's actually extracting an archive in order to let a user know they may want to call `datalad clean` afterwards. This should make the presence of an extraction cache less surprising to users and make them aware of the `clean` command.

The issue leading to this PR and the discussion therein is #5519, although this is not actually solving it. However, at least this is an improvement of the UX aspect of the issue.

Closes #5834

### CODE REVIEW

**Comment by bpoldrack** [comment]
Note, that this is for now replacing PR #5730, as an actual solution turns out to be more of an effort than I can currently afford. So, not closing #5519 really, but at least let the user know what's going on and how to remove that cache.

@mih : Guess, with that the issue can be regarded as addressed within context of https://github.com/datalad/datalad/projects/13, since the "real solution" is actually significantly more than a "simple UX issue".

**Comment by bpoldrack** [comment]
Looking into more reporting on this. Thus proposing `--info` for `clean`, which in turn could be called by `status`.
Proposal in https://github.com/datalad/datalad/pull/5738/commits/73d97d47286f133bea7f11197522941ed11f5b24. WDYT, @datalad/developers ?

Yes - needs tests. I'm aware. :)

**Review by mih — COMMENTED** [review]
Thx, I think linking this into `status` is a reasonable way to give this aspect the visibility it needs in order to avoid accumulation of cruft.

That being said: I personally see the reporting on cleanable locations not as a core function of `status`, hence I do not think it should actually yield such results (unconditionally). A programmatic user that is concerned with cleaning things up, can, will, and has to make a dedicated call to `clean`. With the present change, `clean` will be called twice for them (once with `--info` and once to actually clean). -- and once for everybody else, unconditionally.

I think it would make sense to move the call to clean into `custom_result_summary_renderer()` and only report on cleanable locations, if there are any. This setup ensures that the cycles necessary are only spent when there is a chance that a (human) consumer catches the notification, and, e.g. not in pipes with `datalad -f json`.

**Comment by bpoldrack** [comment]
> I think it would make sense to move the call to clean into custom_result_summary_renderer() and only report on cleanable locations, if there are any. 

Makes sense to me. Thx.

**Comment by bpoldrack** [comment]
> only report on cleanable locations, if there are any.

FTR: That was the case already.

However, changed to what you suggested, I think, @mih. This now includes a commit to pass original kwargs to custom summary renderers (as we already do for custom result renderers).

File: datalad/core/local/status.py
Comment (mih) [inline datalad/core/local/status.py:504]: If for whatever reason the assertion is wrong, it will crash the process. I think this is a bit extreme for a convenience report. Maybe just `return`?

File: datalad/interface/clean.py
Comment (mih) [inline datalad/interface/clean.py:73]: Thinking about this more, I feel that adopting an already established pattern could have benefits here. `run` uses `run [dry-run]` as action specification in result records. It also uses `--dry-run <mode>` as an argument, not `--info`. I cannot immediately come up with a scenario for more than a boolean flag, but it would still make sense to adopt the naming from my POV. Ultimately it is pretty much a dry-run.

File: datalad/interface/clean.py
Comment (mih) [inline datalad/interface/clean.py:81]: If `--dry-run` is adopted, this first half of the doc change can be removed again, I think.

File: datalad/interface/utils.py
Comment (mih) [inline datalad/interface/utils.py:453]: This change breaks the API. I could not find an actual case where this matters (across extensions), but still worth pointing out. The unification with `custom_result_renderer()` makes sense to me in general.

File: datalad/core/local/status.py
Comment (mih) [inline datalad/core/local/status.py:510]: I admittedly did not think about recursion until now. But maybe we should:

```
/ds/hcp/super (git)-[master] % datalad clean -r 
action summary:
  clean (notneeded: 18000)
datalad clean -r  22.20s user 13.67s system 22% cpu 2:38.55 total
```

I think this calls for some kind of switch/config setting, or we decide that switching the renderer to something else is good enough. However, worth acknowledging explicitly one way or another.

File: datalad/core/local/status.py
Comment (mih) [inline datalad/core/local/status.py:518]: We should have a style guide, as each time I am unsure what tone to use. Passive, active, imperative?

What about simply `Temporary data can be removed with DataLad's 'clean' command`?

I personally think these messages should be as concise and interpretation independent as possible.

File: datalad/customremotes/archives.py
Comment (mih) [inline datalad/customremotes/archives.py:403]:
```suggestion
                    self.info("%s special remote created an extraction cache "
                              "under %s. Remove it with DataLad's 'clean' command "
                              "to save disk space." %
```

There can be performance loss when it is removed. I would prefer concrete advice when to do what, rather than a more nebulous statement.

**Review by mih — COMMENTED** [review]
As always, the more one thinks, the more thoughts come out ;-)

Below are a bunch of additional comments. Also this PR has already shed some light for me personally. I am afraid, this kind of output will be with us frequently:

```
% datalad clone https://github.com/psychoinformatics-de/studyforrest-data-structural
install(ok): /tmp/studyforrest-data-structural (dataset)
% cd studyforrest-data-structural
studyforrest-data-structural (git)-[master] % datalad
studyforrest-data-structural (git)-[master] % datalad get sub-01/anat/sub-01_T1w.nii.gz
get(ok): sub-01/anat/sub-01_T1w.nii.gz (file) [from mddatasrc...]                                                          
studyforrest-data-structural (git)-[master] % datalad status
.git/annex/transfer: Discovered 1 annex temporary transfer directory: download
Note: You can remove temporary files anytime by using datalad's 'clean' command.
```

File: datalad/core/local/status.py
Comment (mih) [reply datalad/core/local/status.py:518]: Git does not seem to use 'hint:' or 'note:' markup:

```
Changes not staged for commit:

[...]

It took 13.57 seconds to enumerate untracked files. 'status -uno'
may speed it up, but you have to be careful not to forget to add
new files yourself (see 'git help status').
```

File: datalad/core/local/status.py
Comment (bpoldrack) [reply datalad/core/local/status.py:504]: True. Actually only wanted to see whether that holds throughout the test battery. Replacement kinda depends on the outcome. Either `return` or failures point to how I got it wrong and can adjust.

File: datalad/interface/clean.py
Comment (bpoldrack) [reply datalad/interface/clean.py:73]: No strong opinion on my end. I'm fine with `--dry-run`.

File: datalad/interface/clean.py
Comment (bpoldrack) [reply datalad/interface/clean.py:81]: Agree.

File: datalad/core/local/status.py
Comment (mih) [reply datalad/core/local/status.py:518]: @adswa used yet another (establish) way to give hints https://github.com/datalad/datalad/pull/5726#issuecomment-860619676

which should declare one as *the one*.

File: datalad/interface/utils.py
Comment (bpoldrack) [reply datalad/interface/utils.py:453]: Sufficient to point it out in the changelog or do you think this actually needs some sort of "technical" deprecation?

File: datalad/core/local/status.py
Comment (bpoldrack) [reply datalad/core/local/status.py:510]: I wondered about that, too. If the reporting should become configurable the "problem" appears to be the default. On the one hand: That hint is kinda important to be there by default - if you know you want to enable it, you probably don't really need it. On the other hand: Being more expensive by default is kinda wrong, too.

File: datalad/core/local/status.py
Comment (bpoldrack) [reply datalad/core/local/status.py:518]: > We should have a style guide, as each time I am unsure what tone to use. Passive, active, imperative?

Agree. I struggle everytime.

File: datalad/customremotes/archives.py
Comment (bpoldrack) [reply datalad/customremotes/archives.py:403]: Yeah, I see that. However, this is the special remote itself. Technically it's kinda wrong to point to datalad commands to begin with + long message would be there all the time. Wondering: We usually call annex with `--json` by now, giving us an info object that we decide to spit out (rather than annex itself). So, we could use that for a short message that we catch in datalad and can potentially decide to enhance or swallow from within a context where we know much more about the situation than within the special remote. Not too clear to me at this point, just thinking about principles how to deal with messaging better in general.

File: datalad/core/local/status.py
Comment (bpoldrack) [reply datalad/core/local/status.py:518]: > @adswa used yet another (established?) way to give hints #5726 (comment)

Not established until merged, I'd say ;-)
However, agree - like the result record enhancement much better. Will use that, too.

File: datalad/core/local/status.py
Comment (yarikoptic) [inline datalad/core/local/status.py:467]: FWIW, looking at the diff I am already confused on how  it was
```
 def custom_result_summary_renderer(results, action_summary):
```
in `push` and just 
```
def custom_result_summary_renderer(results):
```
here was supposed to work (different signatures), and with addition of `**kwargs` (exemplified by absent docstrings describing expected parameters) it just makes them even more of "dark magic" to figure out what they expect to be passed into them.

If we are to make those signatures "flexible", as one renderer might want `recursive` and another not, may be we should at least aim to list expected kwargs listed explicitly, and have some `**_` or alike to swallow those it doesn't care about but which might be passed, with `**_` vs `**kwargs` to hint that "I do not care about other options".

File: datalad/core/local/status.py
Comment (yarikoptic) [inline datalad/core/local/status.py:508]: I really don't like such an approach:
- "renderer" which is also a "do house-keeping" action, possibly even failing along its operation, is IMHO "too much"
- `status` operation could have been limited to some subdatasets, and how this would go through the entire hierarchy of 1000s I have? thank you not
- in my view `status` should be a read-only operation, and cleanup shouldn't be its side-effect
- despite it might be a nice thing to have for some cases, but clearly not for all (e.g. I might not like to cleanup anything).

Comment (yarikoptic) [reply datalad/core/local/status.py:508]: d'oh -- ok, blind me now mentioned this new `info=True` so operation would be a read-only, crossed out some items ;-).  Still not sure about the last one wondering if it is wise to add more extra run time on top of already not that speedy `status` when operating on heavy hierarchies.

Comment (bpoldrack) [reply datalad/core/local/status.py:508]: > status operation could have been limited to some subdatasets, and now this would go through the entire hierarchy of 1000s I have?

You mean it should consider `path` passed into `status` in addition to `recursion_limit`, @yarikoptic ?

Comment (bpoldrack) [reply datalad/core/local/status.py:508]: Re "extra run time":

Consider responding to https://github.com/datalad/datalad/pull/5738#discussion_r650906698? ;-)

Comment (yarikoptic) [reply datalad/core/local/status.py:508]: > You mean it should consider `path` passed into `status` in addition to `recursion_limit`, @yarikoptic ?

well, it sounds like that -- or may be any of those extra kwargs aren't needed?  may be paths for relevant subdatasets could be taken from the `results` records passed into `status`? then `clean` would be checked on those datasets (without any recursion)

Comment (bpoldrack) [reply datalad/core/local/status.py:508]: >  may be paths for relevant subdatasets could be taken from the results records passed into status? 

True. Will do that. Makes the signature change independent from this PR (still think a unification w/ result renderer makes sense).

Comment (yarikoptic) [reply datalad/core/local/status.py:508]: continuing on "renderer" taking on new tasks, I wondered: wouldn't it be better to have `status` code to actually do the clean test internally and just enhancing the returned/yielded dataset status record with e.g. extra field
```
"extra_statuses": {
  "clean": [ ... records from clean  ... ]
}
```
or alike, which we could later enhance possibly with other "hints" or information (aren't we getting more info from `git annex info` which might just get here into `annex-info` ), and then leave it to renderer to just take care about rendering them (instead of running those extra checks)?

File: datalad/core/local/status.py
Comment (adswa) [reply datalad/core/local/status.py:518]: I "established" it in #4338 I think, but simply copied Git ;-)

Comment (bpoldrack) [reply datalad/core/local/status.py:518]: Huh. I forgot that. :)

File: datalad/core/local/status.py
Comment (bpoldrack) [reply datalad/core/local/status.py:508]: @yarikoptic  Hm. Two things speak against that from my POV:
1. After @mih chimed in, I explicitly changed from yielding anything about that to only call `clean --info` from within the summary renderer. See https://github.com/datalad/datalad/pull/5738#pullrequestreview-682620241. I think the argument makes sense.
2. There are clean results coming from a direct invocation of `clean`. I think it's strange that - when called from within `status` - you'd have the info in an extra field of the status-result, while when calling directly you get it as `clean-info` results. That seems inconsistent with how we deal with results and such an approach would generally make it harder to act upon results via hooks - you'd need know all the forms in which it can appear. So, I think if it's something that genuinely yielded from another command it should stay that way. Otherwise I'm fine with extra fields.

File: datalad/core/local/status.py
Comment (bpoldrack) [reply datalad/core/local/status.py:467]: I don't think listing expected arguments in a docstring (of that function) makes a lot of sense. Those functions are called from only one internal place. Furthermore it's not a bout a specific function but all of them. It's an interface - "what do I need to implement". So, may be the renderers should just become classes derived from an abstract class that defines the signatures. That would be a place to document it ... WDYT?

Comment (yarikoptic) [reply datalad/core/local/status.py:467]: recording my thoughts from our discussion during weekly meeting: such interface(s) should probably established/documented within `Interface` class itself.  But my comment (if still to be used) on `**kwargs` vs `**_` stands .  But defining interfaces etc probably shouldn't be a part of this PR.

Also, this `**kwargs` here is really the `**allkwargs` (from https://github.com/datalad/datalad/pull/5738/files#diff-37fba60e98afe87a28f73afa6b5bbc3410bad59901363e26256da5584cec265cR331) which, if I got it right, contains all arguments passed into the call.  
By passing it like it is done in this PR via `**kwargs`, would forbid to have `results` kwarg for an `Interface` `__call__`, and makes it really unclear from just looking at the code -- what are those (hence instigated my comment on docstrings, and discussion on interfaces).

I think a better strategy would be to not `**` them and pass as a dedicated `call_kwargs` argument.  This would then:
- not shadow `results` or any other argument to renderers we ever decide to add
- make it explicit on what is that thing we are passing in

File: datalad/interface/utils.py
Comment (yarikoptic) [reply datalad/interface/utils.py:453]: Do we have  a test which would go through all `Interface`'s and test that invocation in command line works as expected? if not -- we wouldn't even notice until we break things up.

Also see my comment [above](https://github.com/datalad/datalad/pull/5738/files#r654449041): I think they should be passed as 

```suggestion
                wrapped_class.custom_result_summary_renderer(*summary_args, call_kwargs=allkwargs)
```
here.

Also, in principle, we could use then `inspect` (and cache the check) if `custom_result_summary_renderer` expects `call_kwargs` or has a swallow all `**` (if that is possible to check -- didn't look): if not --  issue a DeprecationWarning to alert that interface is changing.

File: datalad/core/local/status.py
Comment (bpoldrack) [reply datalad/core/local/status.py]: Changed to just issue a debug log message and proceed with the first discovered refds (which now is only needed to retrieve a config).

File: datalad/interface/clean.py
Comment (bpoldrack) [reply datalad/interface/clean.py]: Done

File: datalad/interface/clean.py
Comment (bpoldrack) [reply datalad/interface/clean.py]: Done

**Comment by bpoldrack** [comment]
@mih @yarikoptic :

Since I didn't see a good solution to performance issues as discussed above, changed to the following: Simply let `status` display a hint to `clean` in the summary by default - no actual probing is performed. Hint can be disabled by a config. This should address the UX aspect, as users become aware that way.

If that's fine with you guys, will edit PR+first post accordingly.
I would not close the original issue, since autocleaning would still be a desirable option, but just take out the UX label and consider this done for the UX project. Agree, @mih?

Otherwise one question remains for me: `datalad.status.hints.clean` may not be ideal. If we want to generalize that approach, we should decide on the config "hierarchy": Are such flags to be collected per command (as I did here), or should we go for a broader collection like `datalad.hints.status-clean`?

**Comment by mih** [comment]
I cannot get behind some of the rational that led to the current state of the PR, but in its present form I would prefer to not merge it. From my POV it adds to the problem of silently growing repositories the problem of output noise:

```
% datalad create sofresh
[INFO   ] Creating a new annex repo at /tmp/sofresh 
[INFO   ] scanning for unlocked files (this may take some time) 
create(ok): /tmp/sofresh (dataset)

% datalad -C sofresh status
nothing to save, working tree clean
Hint: Run 'datalad clean [--dry-run]' to find and remove temporary files used by datalad and git-annex.
Disable 'datalad.status.hints.clean' to not display this message anymore.
```

If we ignore that the hint is visually masking the important information, users will see this unconditionally every time they execute status (which is arguably one, if not the, most frequently executed command). They will have learned to ignore it (or disable it), before they actually get into the situation that `datalad clean --dry-run` would report anything cleanable.

Given that the PR does not longer address the original issue, I'd say that the modification of `status` should be reverted, and only the updates to `clean` be kept. I wanted to look into how the messaging of the new archive remote behaves, but I tripped over https://github.com/datalad/datalad/issues/5800

**Comment by bpoldrack** [comment]
So, just `clean --dry-run` and a message from `datalad-archives` special remote when it creates an extraction cache.
The latter looks like this ATM:

```
❱ datalad get .
[INFO   ] To obtain some keys we need to fetch an archive of size 2.6 kB                                                                         
[INFO   ] datalad-archives special remote is using an extraction cache under /tmp/clone/.git/datalad/tmp/archives/34d49c45cd. Remove it with DataLad's 'clean' command to save disk space. 
get(ok): .ssh/id_appveyor_rsa (file) [from datalad-archives...]
get(ok): .ssh/id_appveyor_rsa.pub (file) [from datalad-archives...]
get(ok): appveyor_key.tar.gz (file) [from origin...]
action summary:
  get (ok: 3)
```

**Review by adswa — COMMENTED** [review]
I haven't looked into this in detail, but came here to say that I like the UX message where it is put now much more than unconditionally within ``status``.
I also like the addition of a ``dry-run`` flag. 

There were a few things I noticed:
When I run it in a random directory without a cache, it just returns nothing. What do you think: would it make sense to report back a "cacheless" state, e.g., "No temporary clutter found"? This behavior would somehow match that of ``status`` (which reports ``nothing to save, working tree clean``) and ``rerun`` (which warn ``No command given`` when nothing would be run). But it would be different from, e.g. ``git clean -dfn``, which also returns nothing if nothing would be cleaned away.

I tried to trigger the hint by downloading and extracting an archive, but didn't see the hint even though there is a temporary download directory left behind:

```
(handbook2) adina@muninn in /tmp/abcd on git:master
❱ datalad download-url --archive -m test https://www.7-zip.org/a/7za920.zip
[INFO   ] Downloading 'https://www.7-zip.org/a/7za920.zip' into '/tmp/abcd/' 
download_url(ok): /tmp/abcd/7za920.zip (file)                                   
add(ok): 7za920.zip (file)                                                      
save(ok): . (dataset)                                                           
[INFO   ] Adding content of the archive /tmp/abcd/7za920.zip into annex AnnexRepo(/tmp/abcd) 
[INFO   ] Initiating special remote datalad-archives 
[INFO   ] Finished adding /tmp/abcd/7za920.zip: Files processed: 4, +annex: 4 
action summary:
  add (ok: 1)
  download_url (ok: 1)
  save (ok: 1)
(handbook2) adina@muninn in /tmp/abcd on git:master
❱ datalad clean --dry-run
.git/annex/transfer: Discovered 1 annex temporary transfer directory: download
```

It may make sense to add it here, too?

I think it would be nice to add ``--dry-run`` to the examples of the command, even if it is kinda straightforward.

**Comment by bpoldrack** [comment]
> I tried to trigger the hint by downloading and extracting an archive, but didn't see the hint even though there is a temporary download directory left behind

That's a different thing, not the extraction cache of the special remote, though.
`.git/annex/transfer/...` is created by git-annex, not by datalad. So, we can't really issue such a hint when it's happening. We'd need to investigate whether annex left something behind, thus running into similar issues as with the original approach here - probing can be costly.


Re no result:
1. FTR - There should be `notneeded` results (which aren't rendered by default)
2. I guess I'm with you - should behave like `status`. Will do.

**Comment by mih** [comment]
> That's a different thing, not the extraction cache of the special remote, though.

I see 

```
% tree -d .git/datalad
.git/datalad
└── tmp
    └── archives
```

after `download-url`, but no report on it, and `clean` not acting on it without `--dry-run`.

**Comment by bpoldrack** [comment]
> I see 

> ```
> % tree -d .git/datalad
> .git/datalad
> └── tmp
>     └── archives
> ```


@mih : I see that, too. However, it's empty. So, there's nothing to clean.

**Comment by bpoldrack** [comment]
@mih , @adswa :

Added a "nothing-to-do" message, added examples, `clean` now deals with empty directories as well and switched to make use of `Path` objects instead.

```
/tmp/some on git:master❱ datalad clean --dry-run
.git/datalad/tmp/archives: Discovered empty temporary archive directory
.git/annex/tmp: Discovered empty temporary annex directory
.git/annex/transfer: Discovered 1 annex temporary transfer directory: download
/tmp/some on git:master❱ datalad clean
clean(ok): .git/datalad/tmp/archives (directory) [Removed empty temporary archive directory]
clean(ok): .git/annex/tmp (directory) [Removed empty temporary annex directory]
clean(ok): .git/annex/transfer (directory) [Removed 1 annex temporary transfer directory: download]
/tmp/some on git:master❱ datalad clean --dry-run
nothing to clean, no temporary locations present.
```

**Comment by adswa** [comment]
I like it! 
But one of your tests fails on Windows:

```
======================================================================
ERROR: datalad.interface.tests.test_clean.test_clean
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python39-x64\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "C:\Python39-x64\lib\site-packages\datalad\tests\utils.py", line 739, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "C:\Python39-x64\lib\site-packages\datalad\interface\tests\test_clean.py", line 55, in test_clean
    res = clean(dataset=ds, return_type='item-or-list',
  File "C:\Python39-x64\lib\site-packages\datalad\interface\utils.py", line 483, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "C:\Python39-x64\lib\site-packages\datalad\interface\utils.py", line 476, in return_func
    results = list(results)
  File "C:\Python39-x64\lib\site-packages\datalad\interface\utils.py", line 396, in generator_func
    for r in _process_results(
  File "C:\Python39-x64\lib\site-packages\datalad\interface\utils.py", line 574, in _process_results
    for res in results:
  File "C:\Python39-x64\lib\site-packages\datalad\interface\clean.py", line 173, in __call__
    rmtree(topdir)
  File "C:\Python39-x64\lib\site-packages\datalad\utils.py", line 476, in rmtree
    path = r'\\?\ '.strip() + path
TypeError: can only concatenate str (not "WindowsPath") to str
````

**Comment by bpoldrack** [comment]
Ah - stupid me. Force-pushed.

**Comment by adswa** [comment]
Cool, works for me locally on Windows now!

**Review by mih — CHANGES_REQUESTED** [review]
I resolved half a dozen conversation on topics that are no longer associated with the changes proposed in here. It would probably have been better to close this PR and start a new one, instead. It is needlessly hard to review these winding stories.

Given that this PR closes #5834 I think a key change is the special remote info message. However, there is no test for it actually showing up in user output, and I cannot make it show manually either. 

```
% datalad create archive
[INFO   ] Creating a new annex repo at /tmp/archive 
[INFO   ] scanning for unlocked files (this may take some time) 
create(ok): /tmp/archive (dataset)
% cd archive
% datalad download-url --archive -m test https://www.7-zip.org/a/7za920.zip
[INFO   ] Downloading 'https://www.7-zip.org/a/7za920.zip' into '/tmp/archive/' 
download_url(ok): /tmp/archive/7za920.zip (file)                                           
add(ok): 7za920.zip (file)                                                                 
save(ok): . (dataset)                                                                      
[INFO   ] Adding content of the archive /tmp/archive/7za920.zip into annex AnnexRepo(/tmp/archive) 
[INFO   ] Initiating special remote datalad-archives 
[INFO   ] Finished adding /tmp/archive/7za920.zip: Files processed: 4, +annex: 4 
action summary:
  add (ok: 1)
  download_url (ok: 1)
  save (ok: 1)
(datalad-dev) mih@meiner /tmp/archive (git)-[master] % git annex info
...
        c04eb54b-4b4e-5755-8436-866b043170fa -- [datalad-archives]
```

Is it not working? Or is it not supposed to be working in this case?

**Comment by bpoldrack** [comment]
> Is it not working? Or is it not supposed to be working in this case?

Not supposed. `download_url` doesn't leave an extraction cache. However, if you drop the extracted content, clone that ds and `get` the content from origin, the cache should be established and the message be issued.

Will add a test.

**Comment by mih** [comment]
> Not supposed. `download_url` doesn't leave an extraction cache. However, if you drop the extracted content, clone that ds and `get` the content from origin, the cache should be established and the message be issued.

Thanks for the clarification. I can confirm that it does not need a clone to see it working:

(continued from above)
```
(datalad-dev) mih@meiner /tmp/archive (git)-[master] % datalad drop license.txt
drop(ok): /tmp/archive/license.txt (file)                                                  
(datalad-dev) mih@meiner /tmp/archive (git)-[master] % datalad get license.txt
[INFO   ] datalad-archives special remote is using an extraction cache under /tmp/archive/.git/datalad/tmp/archives/5ca17613fa. Remove it with DataLad's 'clean' command to save disk space. 
get(ok): license.txt (file) [from datalad-archives...]
```

**Comment by bpoldrack** [comment]
> I can confirm that it does not need a clone to see it working:

Yeah, just reread my own comment and noticed the cloning was nonsense. 

Pushed a test enhancement.

**Review by mih — APPROVED** [review]
Thx! From my POV this is ready.

**Pull request merged**
